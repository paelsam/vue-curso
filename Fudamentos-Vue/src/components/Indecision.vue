<template>
    <img v-if="sourceImage" :src="sourceImage" alt="bg">
    <div class="bg-dark"></div>

    <div class="indecision-container">
        <input
          type="text"
          placeholder="Hazme una pregunta"
          v-model="question">
        <p>Recuerda terminar con un signo de interrogaciÃ³n (?) ðŸ˜‰</p>
    </div>

    <div v-if="isValidQuestion" class="answer-container">
        <h2>{{ question }}</h2>
        <h1>{{ answer }}</h1>
    </div>
</template>

<script>
export default {
    data() {
        return {
            question: null,
            answer: null,
            sourceImage: null,
            isValidQuestion: false
        }
    },
    methods: {
        async getAnswerAndSourceImage() {
            this.answer = 'Thinking...'
            try {
                const { answer, image } = await fetch('https://yesno.wtf/api')
                                .then( data => data.json() );
                this.answer = (answer.toUpperCase() === 'YES') ? 'SÃ­!' : 'No!';
                this.sourceImage = image;
            } catch(err) {
                throw new Error( err );
            }
        }
    },
    watch: {
        question( value, oldValue )  {
            this.isValidQuestion = false

            console.log({ value });

            if ( !value.includes('?') ) return 
            this.isValidQuestion = true
            // !TODO: Realizar peticiÃ³n http (Realizado)
            this.getAnswerAndSourceImage()
        } 
    }

}

</script>

<style scoped>
img, .bg-dark {
    height: 100vh;
    left: 0px;
    max-height: 100%;
    max-width: 100%;
    position: fixed;
    top: 0px;
    width: 100vw;
}

.bg-dark {
    background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    position: relative;
    z-index: 99;
}

.answer-container {
    z-index: 99;
    text-align: center;
    position: relative;
}

input {
    width: 250px;
    padding: 10px 15px;
    border-radius: 5px;
    border: none;
}

input:focus {
    outline: none;
}

p {
    color: white;
    font-size: 20px;
    margin-top: 0px;
}

h1,
h2 {
    color: white;
}

h2 {
    margin-top: 150px;
}
</style>