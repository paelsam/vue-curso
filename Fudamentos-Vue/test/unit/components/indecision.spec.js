import Indecision from '../../../src/components/Indecision.vue'
import { shallowMount } from "@vue/test-utils";
import { test, describe, beforeEach, expect, vitest, vi } from "vitest";


describe('Indecision Component', () => {

    // Preparando el componente para las pruebas
    let wrapper
    // Spy: están pendientes de cualquier caso (clg)
    let clgSpy


    // Metodo que se ejecuta en cada test
    beforeEach(() => {
        wrapper = shallowMount(Indecision )

        clgSpy = vitest.spyOn( console, 'log' )

        vitest.clearAllMocks()
    })

    /* Test space */

    test('Haciendo un snapshop al componente Indecision', () => {
        expect( wrapper.html() ).toMatchSnapshot();
    })

    test('Escribir en el input no debe disparar nada (clg)', async () => {
        
        const getAnswerSpy = vitest.spyOn( wrapper.vm, 'getAnswerAndSourceImage' )
        
        const $input = wrapper.find('input')
        await $input.setValue('Hola Mundo')
        
        expect( clgSpy ).toHaveBeenCalled()
        expect( getAnswerSpy ).not.toHaveBeenCalled(0)
        
        // Accedemos a la instanciación del coponente de vue
        // console.log( wrapper.vm );
    })

    test('Escribir el símbolo de "?" debe disparar el getAnswer', async () => {
        // Mock: Simulación de una petición http
        const getAnswerAndSourceImageSpy = vitest.spyOn( wrapper.vm, 'getAnswerAndSourceImage' )
        
        const $input = wrapper.find('input')
        await $input.setValue('Hola Mundo?')
        
        expect( getAnswerAndSourceImageSpy ).toHaveBeenCalled( )
    })

    test('Pruebas en getAnswer', async () => {


    })

    test('Pruebas en getAnswer - Fallo en el API', () => {

    })

})