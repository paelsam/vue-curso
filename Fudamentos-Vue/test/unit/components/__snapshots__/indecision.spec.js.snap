// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Indecision Component > Haciendo un snapshop al componente Indecision 1`] = `
"<!--v-if-->
<div data-v-4a8fca50=\\"\\" class=\\"bg-dark\\"></div>
<div data-v-4a8fca50=\\"\\" class=\\"indecision-container\\"><input data-v-4a8fca50=\\"\\" type=\\"text\\" placeholder=\\"Hazme una pregunta\\">
  <p data-v-4a8fca50=\\"\\">Recuerda terminar con un signo de interrogaciÃ³n (?) ğŸ˜‰</p>
</div>
<!--v-if-->"
`;
